Fixes an issue with hosts having an IPv6 address on localhost

akram opened the pull request:
- When 'getent hosts localhost' returns '::1' the creation of the listener fails because of the port parsing which uses ":" as a separator
- Use of net.SplitHostPort() to do the job
- Adding unit tests to ensure that the creation succeeds
- On docker.go: adds a test on the presence the socat command which was failing silenty if not installed

akram commented:
Hi guys,

this PR fixes an issue on host that have its /etc/hosts having a ::1 address mapping to localhost, and on which name resolution to localhost first resolves to the IPv6 address, which seems to be the case for fedora and arch hosts. Probably more.

To check if it the the case for your distribution run the following command:
```
getent hosts localhost
```
if it returns ::1, the kubectl port-forward will crash with the following error message:

```
W0331 09:47:04.409552   21488 portforward.go:165] Unable to listen on port {5001 5000}: Error parsing local part: strconv.ParseUint: parsing "": invalid syntax
```
This happens because of the returned address address from the Listen command used to bind the local port on "localhost" is then parsed by looking for a ":" char, which fails in case of the localhost is resolved to "::1"

@ncdc and others, could you please review ?

akram commented:
I am the author of these commits

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+	stopChan := make(chan struct{}, 1)
+	pf, err := New(&client.Request{}, &client.Config{}, []string{"15000", "15000"}, stopChan)
+	if err != nil {
+		t.Logf("Cannot create PortForwarder %v: %v", pf, err)
```
This should be `t.Fatalf` as the remainder of the test code won't run without a PortForwarder

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+		Protocol                  string
+		Ports                     ForwardedPort
+		ShouldRaiseError          bool
+		ExecpectedListenerAddress string
```
type: Expected (same for next line)

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+	testCases := []struct {
+		Hostname                  string
+		Protocol                  string
+		Ports                     ForwardedPort
```
I don't see any need to vary the ports in the test cases. I'd recommend hardcoding something inside the for loop instead.

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+		},
+
+		{
+			Hostname:         "www.google.com",
```
I don't really think you need this test case

akram replied:
This was one of the concerns of this PR.
Even even those tests are not 100% relevant for the current scenario, that helps to understand the expected behaviour.

More over, there are stil some stickyness to TCP4 in the docker.go implementation, once that we definitely want to get rid of this, and be able to handled tcp4 and tcp6 correctly, these tests would help.

If someone else agrees, I can remove these tests.

ncdc replied:
The expected behavior is that we listen on localhost, not on google.com.

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+			ShouldRaiseError: true,
+		},
+		{
+			Hostname:         "1.1.1.1",
```
I don't really think you need this test case

akram replied:
same comment as above :)

ncdc replied:
I still disagree with including them. We're hardcoding localhost. I would actually modify `getListener` to not take in a hostname.

akram replied:
This is the purpose of this PR: To put in evidence the different behaviours of the net.Listen method depending on name resolution in the host.
These different behaviours are the cause of this issue in conjonction with the way we were parsing the resulting bind address.

ncdc replied:
It's to make sure that we can get the port from "localhost:$port", where $port may be 0, in which case the listener's Addr will contain an arbitrary ephemeral port. Please remove tests for anything other than localhost an ::1.

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+		errorRaised := err != nil
+
+		if testCase.ShouldRaiseError != errorRaised {
+			t.Fatalf("Test case #%d failed: Data %v an error has been raised(%t) where it should not (or reciprocally): %v", i, testCase, testCase.ShouldRaiseError, err)
```
Do `t.Errorf` and `continue`

akram replied:
will do that, commit is pending

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+		if testCase.ShouldRaiseError != errorRaised {
+			t.Fatalf("Test case #%d failed: Data %v an error has been raised(%t) where it should not (or reciprocally): %v", i, testCase, testCase.ShouldRaiseError, err)
+		}
+		if listener != nil {
```
Convert to `if listener == nil { t.Errorf(); continue }` and remove the else

akram replied:
will do that, commit is pending

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+			t.Logf("Asked a %s forward for: %s:%v, got listener %s:%s", testCase.Protocol, testCase.Hostname, testCase.Ports, host, port)
+			if err != nil {
+				if testCase.ExpectedListenerAddress != host {
+					t.Fatalf("Listener does not listen on exepected address: asked %v got %v", testCase.ExpectedListenerAddress, host)
```
Errorf

Add test case # in output

akram replied:
will do that, commit is pending

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,102 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+					t.Fatalf("Listener does not listen on exepected address: asked %v got %v", testCase.ExpectedListenerAddress, host)
+				}
+				if testCase.ExpectedListenerPort != port {
+					t.Fatalf("Listener does not listen on exepected port: asked %v got %v", testCase.ExpectedListenerPort, port)
```
Errorf

Add test case # in output

akram replied:
will do that, commit is pending

ncdc commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -316,7 +316,11 @@ func (d *dockerContainerCommandRunner) PortForward(podInfraContainerID string, p
+	// TODO what if the host doesn't have it???
+	path, lookupErr := exec.LookPath("socat")
+	if lookupErr != nil {
+		return fmt.Errorf("The socat command cannot be find in PATH=%s.", path)
```
Just say "Port forwarding is unavailable: unable to locate socat"

akram replied:
More information isn't better than no information?

ncdc replied:
I can't remember if this error is returned to the client, but if so, I don't think it's appropriate to tell the end user what the PATH is on the kubelet. You could certainly log this so it shows up in the kubelet's log file, but I don't think the end user should get the PATH info.

ncdc replied:
And it's probably better not to mention socat either

akram replied:
Agree with the PATH information not necessarily useful.

Regarding not mentioning socat: That would be very obscure. Not being able to know the cause of the  "Port forwarding is unavailable" is very frustrating. Why hiding it ?

ncdc replied:
Reduce potential attack vectors from bad actors.

akram replied:
Security by obfuscation cannot be an argument....certainly in open source

ncdc replied:
Alright

ncdc commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -316,7 +316,11 @@ func (d *dockerContainerCommandRunner) PortForward(podInfraContainerID string, p
 
 	containerPid := container.State.Pid
-	// TODO use exec.LookPath for socat / what if the host doesn't have it???
+	// TODO what if the host doesn't have it???
```
Remove the TODO

akram replied:
the question about what to do is still valid, for this moment, it is only failing with an error.
If you think that this is really sufficient, I can remove it.

ncdc replied:
I think the most appropriate way to handle this is to have the kubelet check for socat and return an error if it's not available when the port forwarding request comes in, before spawning listeners and so on. But that's not implemented now, and doesn't need to be part of this PR.

akram replied:
agree
let's keep this for another PR for this specific point
I was only touching this part to remove frustration of not knowing the cause of the error.

akram commented:
@ncdc thanks, for reviewing. I implemented your remarks.

 Team (@smarterclayton), could you please let me know your remarks?

For the moment, I will commit on top, to of course eventually squash and rebase the whole PR.

ncdc commented on pkg/client/portforward/portforward.go:
```diff
@@ -205,6 +197,27 @@ func (pf *PortForwarder) listenOnPort(port *ForwardedPort) error {
 
+// getListener creates a listener on the interface targeted by the given hostname on the given port with
+// the given protocol. protocol is in net.Listen style which basically admits values like tcp, tcp4, tcp6
+func (pf *PortForwarder) getListener(protocol string, hostname string, port *ForwardedPort) (net.Listener, error) {
```
remove `hostname`

ncdc commented on pkg/client/portforward/portforward.go:
```diff
@@ -205,6 +197,27 @@ func (pf *PortForwarder) listenOnPort(port *ForwardedPort) error {
+// getListener creates a listener on the interface targeted by the given hostname on the given port with
+// the given protocol. protocol is in net.Listen style which basically admits values like tcp, tcp4, tcp6
+func (pf *PortForwarder) getListener(protocol string, hostname string, port *ForwardedPort) (net.Listener, error) {
+	listener, err := net.Listen(protocol, fmt.Sprintf("%s:%d", hostname, port.Local))
```
remove `hostname` and hardcode `localhost`

ncdc replied:
Oh I know why you're including it...

akram replied:
So I can keep it? right?

ncdc replied:
Yes

akram commented:
@ncdc do you see other points ?

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,86 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+	testCases := []struct {
+		Hostname                string
+		Protocol                string
+		Ports                   ForwardedPort
```
I still don't think you need this - just hardcode something at the beginning of the for loop as these values don't need to vary.

akram replied:
agree
I removed initialisation of pf. It is already tested elsewhere.
commit pending

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,86 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+			t.Errorf("Test case #%d failed: Data %v an error has been raised(%t) where it should not (or reciprocally): %v", i, testCase, testCase.ShouldRaiseError, err)
+			continue
+		}
+		if listener != nil {
```
```go
if listener == nil {
  t.Errorf("...")
  continue
}

host, port, err ...
```

akram replied:
Getting a nil listener is not an necessarely an erroneous test case.
We agree that the test case is considered in error if an error has not been raised where it should and vice-versa
And, when an error is not raised, the listener is not nil, and so, we have other tests then.

Or I misunderstood that. Failing when listener == nil definitely broke the test

ncdc replied:
If the test case should raise an error, and it did not, it will be caught by the previous `if` and then `continue`d. If the test case should not raise an error, the listener therefore should not be nil. If it is nil, this is an error.

akram replied:
not exactly my comprehension:
OK, I got it, but, would the following be more accurate: 
If the test case dit not raise an error, the listener therefore should not be nil

I will commit this

ncdc replied:
Whatever works for understanding it :smile: 

Basically if you don't expect an error, and you get a nil listener, that is a problem.

ncdc commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -316,7 +316,11 @@ func (d *dockerContainerCommandRunner) PortForward(podInfraContainerID string, p
+	// TODO what if the host doesn't have it???
+	_, lookupErr := exec.LookPath("socat")
+	if lookupErr != nil {
+		return fmt.Errorf("The socat command cannot be find.")
```
Unable to do port forwarding: socat not found.

akram replied:
done, commit is pending

bprashanth commented:
@ncdc i can't assign to you, so I'm assigning to Derek instead

derekwaynecarr commented:
I am fine being @ncdc proxy on this, will wait for his LGTM

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,85 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+	testCases := []struct {
+		Hostname                string
+		Protocol                string
+		Ports                   ForwardedPort
```
Please remove this from the testCase struct.

akram replied:
OK
commit pending

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,85 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+		Ports                   ForwardedPort
+		ShouldRaiseError        bool
+		ExpectedListenerAddress string
+		ExpectedListenerPort    string
```
Please remove this from the testCase struct. You can hardcode a port below in the loop.

akram replied:
OK commit pending

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,86 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+		listener, err := pf.getListener(testCase.Protocol, testCase.Hostname, &ForwardedPort{12345, 12345})
+		errorRaised := err != nil
+
+		if testCase.ShouldRaiseError != errorRaised {
```
I think this is cleaner

```go
if testCase.ShouldRaiseError != errorRaised {
	t.Errorf("Test case #%d failed: Data %v an error has been raised(%t) where it should not (or reciprocally): %v", i, testCase, testCase.ShouldRaiseError, err)
	continue
}
if errorRaised {
  continue
}

if listener == nil {
	t.Errorf("Test case #%d did not raised an error (%t) but failed in initializing listener", i, err)
	continue
}

host, port, _ := net.SplitHostPort(listener.Addr().String())
```

akram replied:
ok,
commit pending

akram replied:
committed

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,88 @@ func (s *fakeUpgradeStream) Headers() http.Header {
 	return http.Header{}
 }
 
+func CloseListener(l net.Listener) {
```
This method isn't used?

akram replied:
oh, yes, sorry
Added it to properly handle closing in Code Review 5, forgot to remove it.
Will amend my last commit

ncdc commented on pkg/client/portforward/portforward_test.go:
```diff
@@ -206,6 +206,88 @@ func (s *fakeUpgradeStream) Headers() http.Header {
+			t.Errorf("Test case #%d failed: Listener does not listen on exepected port: asked %v got %v", i, expectedListenerPort, port)
+
+		}
+		listener.Close()
```
Sorry, I just noticed this - can you move this up to right before you call SplitHostPort and call it via defer?

ncdc replied:
Oh wait, that won't work because the same port is being reused.

ncdc commented:
LGTM

@smarterclayton PTAL

derekwaynecarr commented:
@akram - can you squash your commits down to 1?

akram commented:
wow please wait, I don't know what happened when squashing my commits

akram commented:
@derekwaynecarr done for squashing

akram committed:
Fixes an issue with hosts having an IPv6 address on localhost
- When 'getent hosts localhost' returns '::1' the creation of the
  listener fails because of the port parsing which uses ":" as a
separator
- Use of net.SplitHostPort() to do the job
- Adding unit tests to ensure that the creation succeeds
- On docker.go: adds a test on the presence the socat command which was
  failing silenty if not installed

- Code Review 1
  - Fixed typo on Expected
  - The UT now fails if the PortForwarder could not be created

- Code Review 2
  - Simplify socat error message
  - Changing t.Fatal to to.Error on unit tests

- Code Review 3
  - Removing useless uses cases in unit tests

- Code Review 4
  - Removing useless initiliasiation of PortForwarder
  - Changing error message

- Code Review 5
  - Simplifying TestCast struct
  - Adding addition test in one test case
  - Closing the listener

- Code Review 6
  - Improving unit test

akram commented:
@derekwaynecarr :ping: squashed and rebased. Travis and shippable also seems ok.
